source/extensions/tracers/xray/localized_sampling.cc

```diff
@@ -0,0 +1,185 @@
+#include "extensions/tracers/xray/localized_sampling.h"
+
+#include "common/http/exception.h"
+
+#include "extensions/tracers/xray/util.h"
+
+#include "rapidjson/document.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace XRay {
+
+constexpr static double DefaultRate = 0.5;
+constexpr static int DefaultFixedTarget = 1;
+constexpr static int SamplingFileVersion = 2;
+constexpr static char VersionJsonKey[] = "version";
+constexpr static char DefaultRuleJsonKey[] = "default";
+constexpr static char FixedTargetJsonKey[] = "fixed_target";
+constexpr static char RateJsonKey[] = "rate";
+constexpr static char CustomRulesJsonKey[] = "rules";
+constexpr static char HostJsonKey[] = "host";
+constexpr static char HttpMethodJsonKey[] = "http_method";
+constexpr static char UrlPathJsonKey[] = "url_path";
+
+// Making the following two functions (fail() and validate()) class-members requires:
+// 1- making LocalizedSamplingManifest inherit from the logger (for no apparent benefit other
+// than saving us from typing Logger::Id::tracing)
+// 2- adding rapidjson header to localized_sampling.h
+// 3- cluttering the interface of class LocalizedSamplingManifest.
+// Moreover, Registry::getLog is a cheap call (array lookup) and happens once per parsing. Once we
+// fail, we don't continue, much better than throwing exception
+void fail(absl::string_view msg) {
```

Comment:
shouldn't these be static functions then?

Reply:
Yes. Thank you.